### DISCUSSION TITLE
Why AVX512 is slower than AVX2 ?

### DISCUSSION THREAD

**Discussion by luoyger** [milvus-io/milvus#14942] [Q&A and General discussion]
Refer inner product computing in milvus, test the inner product by AVX2 and AVX512,  method `vecInnerProduct256` is faster than method `vecInnerProductAvx512` and `vecInnerProductAvx512_2` .
Is there any mistake for this test ?

```
include <immintrin.h>
// AVX2method1
float vecInnerProduct256(const float* x, const float* y, size_t d) {

    __m256 msum1 = _mm256_setzero_ps();
    while (d >= 8) {
        __m256 mx = _mm256_loadu_ps(x);
        x += 8;
        __m256 my = _mm256_loadu_ps(y);
        y += 8;
        msum1 = _mm256_add_ps(msum1, _mm256_mul_ps(mx, my));
        d -= 8;
    }

    __m128 msum2 = _mm256_extractf128_ps(msum1, 1);
    msum2 = _mm_add_ps(msum2, _mm256_extractf128_ps(msum1, 0));

    msum2 = _mm_hadd_ps(msum2, msum2);
    msum2 = _mm_hadd_ps(msum2, msum2);
    return _mm_cvtss_f32(msum2);
}

// AVX512 method1
float vecInnerProductAvx512(const float* x, const float* y, size_t d) {
    __m512 msum1 = _mm512_setzero_ps();

    while (d >= 16) {
        __m512 mx = _mm512_loadu_ps(x);
        x += 16;
        __m512 my = _mm512_loadu_ps(y);
        y += 16;
        msum1 = _mm512_add_ps(msum1, _mm512_mul_ps(mx, my));
        d -= 16;
    }

    __m256 msum2 = _mm512_extractf32x8_ps(msum1, 1);
    msum2 = _mm256_add_ps(msum2, _mm512_extractf32x8_ps(msum1, 0));

    __m128 msum3 = _mm256_extractf128_ps(msum2, 1);
    msum3 = _mm_add_ps(msum3, _mm256_extractf128_ps(msum2, 0));

    msum3 = _mm_hadd_ps(msum3, msum3);
    msum3 = _mm_hadd_ps(msum3, msum3);
    return _mm_cvtss_f32(msum3);
}

// AVX512 method2
float vecInnerProductAvx512_2(const float* x, const float* y, size_t d) {
    __m512 msum1 = _mm512_setzero_ps();

    while (d >= 16) {
        __m512 mx = _mm512_loadu_ps(x);
        x += 16;
        __m512 my = _mm512_loadu_ps(y);
        y += 16;
        msum1 = _mm512_fmadd_ps(mx, my, msum1);
        d -= 16;
    }
   return _mm512_reduce_add_ps(msum1);
}
```

Compile command
`clang -O3 -mavx512f -mavx512dq -mfma -masm=intel -mno-red-zone -mstackrealign -mllvm -inline-threshold=1000 -fno-asynchronous-unwind-tables -fno-exceptions -fno-rtti avx_test.c -o avx_test
`

CPU info (AVX512 is enabled)
`fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology eagerfpu pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch fsgsbase bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 arat`

**Comment by yhmo** [comment]
Well, AVX512 is not always faster than AVX2, I have no idea about the answer, please refer to these links:
https://www.zhihu.com/question/406517759
https://www.reddit.com/r/intel/comments/71tm03/how_much_faster_is_avx512_vs_avx256/
https://community.intel.com/t5/Software-Tuning-Performance/AVX512-slower-than-AVX2-with-Intel-MKL-dgemm-on-Intel-Gold-5118/td-p/1135951

**Comment by JackLCL** [comment]
@luoyger 
This is a question you may need to ask at Intel's website.

**Comment by cydrain** [comment]
Here is another article to explain why AVX512 sometimes is slower:

Why Skylake CPUs Are Sometimes 50% Slower – How Intel Has Broken Existing Code
https://aloiskraus.wordpress.com/2018/06/16/why-skylakex-cpus-are-sometimes-50-slower-how-intel-has-broken-existing-code/
